Upgrades Custom Messages to Hooks and ShipInit

leggettc18 opened the pull request:
This PR is my effort to convert our old Custom Message systems to use the OnTextLoad hook that I backported from 2Ship, and ShipInit to replace our previous methods for displaying Custom Messages. Some articles, colors, and icons got lost in translation that I know of, so it's not 100% done yet. However these are all visual, so playtesters are safe to go ahead and try this build out to see if they can find any bugs like text popping up in the wrong places or hooks not getting registered properly.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/5144782628.zip)
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/5144821100.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/5144991780.zip)
<!--- section:artifacts:end -->

leggettc18 committed:
Defines the hook for OnTextLoad

leggettc18 committed:
Defines the hook for OnTextLoad

leggettc18 committed:
Intercepts message loading with hooks

leggettc18 committed:
Intercepts message loading with hooks

leggettc18 committed:
Adds file to handle the CustomMessage creation for items.

leggettc18 committed:
Adds file to handle the CustomMessage creation for items.

leggettc18 committed:
Handles Ice Traps, Triforce Pieces, and Custom Items

leggettc18 committed:
Handles Ice Traps, Triforce Pieces, and Custom Items

leggettc18 committed:
Handle maps, compasses, and keys

leggettc18 committed:
Handle maps, compasses, and keys

leggettc18 committed:
Set up actual loading of messages into memory

Only for the items, other things previously covered by CustomMessage_Retrieve if exists have not yet been covered as of this commit and will either load their vanilla text or crash.

leggettc18 committed:
Set up actual loading of messages into memory

Only for the items, other things previously covered by CustomMessage_Retrieve if exists have not yet been covered as of this commit and will either load their vanilla text or crash.

leggettc18 committed:
Remove now-unused code

leggettc18 committed:
Remove now-unused code

leggettc18 committed:
Converts gossip stone hints to hook

leggettc18 committed:
Converts gossip stone hints to hook

leggettc18 committed:
Moved into "Messages" folder

leggettc18 committed:
Moved into "Messages" folder

leggettc18 committed:
Handle one-off merchant messages

leggettc18 committed:
Handle one-off merchant messages

leggettc18 committed:
Convert scrubs and shop text and remove now-unused code

leggettc18 committed:
Convert scrubs and shop text and remove now-unused code

leggettc18 committed:
Moves comment about shops

leggettc18 committed:
Moves comment about shops

leggettc18 committed:
Convert Sheik and Ganondorf hint text

leggettc18 committed:
Convert Sheik and Ganondorf hint text

leggettc18 committed:
Convert skulltula people messages

leggettc18 committed:
Convert skulltula people messages

leggettc18 committed:
Convert more static hints

Specifically Dampe, Greg, Warp Songs, Frogs, Loach, Fishing Pole, and Saria

leggettc18 committed:
Convert more static hints

Specifically Dampe, Greg, Warp Songs, Frogs, Loach, Fishing Pole, and Saria

leggettc18 committed:
Remove now-unused Sheik hint functions

leggettc18 committed:
Remove now-unused Sheik hint functions

leggettc18 committed:
Convert Biggoron Hint

leggettc18 committed:
Convert Biggoron Hint

leggettc18 committed:
Convert Big Poes hint

leggettc18 committed:
Convert Big Poes hint

leggettc18 committed:
Adds TODO comment about breaking these up later.

leggettc18 committed:
Adds TODO comment about breaking these up later.

leggettc18 committed:
Convert Anju hint dialogue

leggettc18 committed:
Convert Anju hint dialogue

leggettc18 committed:
Convert Malon hint

leggettc18 committed:
Convert Malon hint

leggettc18 committed:
Convert Horseback Archery hints

leggettc18 committed:
Convert Horseback Archery hints

leggettc18 committed:
Convert Mask Shop SIgn hint

This is the last of the static hints

leggettc18 committed:
Convert Mask Shop SIgn hint

This is the last of the static hints

leggettc18 committed:
Add some documentation comments

leggettc18 committed:
Add some documentation comments

leggettc18 committed:
Handles previously missed Carpet Guy can't buy message.

leggettc18 committed:
Handles previously missed Carpet Guy can't buy message.

leggettc18 committed:
Removes some more unused stuff

leggettc18 committed:
Removes some more unused stuff

leggettc18 committed:
Convert Lake Hylia Switch related text

leggettc18 committed:
Convert Lake Hylia Switch related text

leggettc18 committed:
Convert Shooting Gallery bow reminder message

leggettc18 committed:
Convert Shooting Gallery bow reminder message

leggettc18 committed:
Removes now-unused stuff for hint messages

leggettc18 committed:
Removes now-unused stuff for hint messages

leggettc18 committed:
Convert random rupee names

leggettc18 committed:
Convert random rupee names

leggettc18 committed:
Convert Rando-Relevant Navi Enhancement

leggettc18 committed:
Convert Rando-Relevant Navi Enhancement

leggettc18 committed:
Convert Random Goron messages

leggettc18 committed:
Convert Random Goron messages

leggettc18 committed:
Removes final randomizer message unused stuff

leggettc18 committed:
Removes final randomizer message unused stuff

leggettc18 committed:
Convert Injecting Skull token counts

leggettc18 committed:
Convert Injecting Skull token counts

leggettc18 committed:
Add in a way to AutoFormat with an item icon.

leggettc18 committed:
Add in a way to AutoFormat with an item icon.

leggettc18 committed:
Fix some errors with skull tokens and apply icon

leggettc18 committed:
Fix some errors with skull tokens and apply icon

leggettc18 committed:
Convert heart container item counts

leggettc18 committed:
Convert heart container item counts

leggettc18 committed:
Convert Inject Item Count for Heart Pieces.

leggettc18 committed:
Convert Inject Item Count for Heart Pieces.

leggettc18 committed:
Port Better Bombchu Shopping

leggettc18 committed:
Port Better Bombchu Shopping

leggettc18 committed:
Convert Market Sneak

leggettc18 committed:
Convert Market Sneak

leggettc18 committed:
Port Quit Fishing At Door and clean up unused stuff

leggettc18 committed:
Port Quit Fishing At Door and clean up unused stuff

leggettc18 committed:
Fix an overlap on Greg with Potion shop guy

leggettc18 committed:
Fix an overlap on Greg with Potion shop guy

leggettc18 committed:
Fixes an incorrect color for the Scrub text

leggettc18 committed:
Fixes an incorrect color for the Scrub text

leggettc18 committed:
Reintroduce missing Mysterious warp song hint

leggettc18 committed:
Reintroduce missing Mysterious warp song hint

leggettc18 committed:
Make the the ShipInitFuncs static

leggettc18 committed:
Make the the ShipInitFuncs static

leggettc18 committed:
One more missed static initFunc

leggettc18 committed:
One more missed static initFunc

leggettc18 committed:
Adds and uses per-item articles for get item messages

leggettc18 committed:
Add spaces to articles

leggettc18 committed:
Fix skulltula hint messages

leggettc18 committed:
Add text colors to items and respect them in most messages

leggettc18 committed:
Clean up now unused skulltula hint storage

leggettc18 committed:
Fix rebase conflicts and port Skulltula Freeze fix

Pepper0ni commented on soh/soh/Enhancements/randomizer/Messages/StaticHints.cpp:
```diff
@@ -0,0 +1,437 @@
+        return;
+    }
+    Rando::Item& item = Rando::StaticData::RetrieveItem(RAND_GET_ITEM_LOC(rc)->GetPlacedRandomizerGet());
+    CustomMessage msg = CustomMessage(
```
Is there a particular reason skulls got ripped out of StaticHintInfoMap and not the other fixed location hints? Or for that matter, why we are splitting up the different ways to process hint text again...

Also, while I can see wanting to replace HintTextKeys sooner or later, it comes across as backwards to remove text from a dedicated repository just to hardcode it somewhere else out of the way and "invisible". I think between this and the conflict in our item name thoughts we need to figure out how we plan on actually storing text in ship in the long term because it's not productive to have us ripping out each others stuff when we both know that and some point everything is going to want change around again and exposing to modding so mod translations can happen.

leggettc18 replied:
I was having a change of opinion about storing the text ahead of time vs just building it when its needed after seeing 2ship's systems in action, but mod translations does throw a bit of a wrench in that admittedly.

I actually was planning on putting this back on draft and changing this again, so that the idea would be that functionality (including hints, but not *only* hints) would be broken up by Actor. So we'd find the actors for the skulltula house people, make an ActorBehavior file for them like 2ship does, the hint text hook register would happen in there, and the hint text is also in there, and if we wanted to change the amount of tokens required to collect their rewards, that would also be in that same file.

What I'm thinking is gonna happen is that when translation mods are a thing, is that the text isn't gonna end up existing in C/CPP files at all anyway, it's gonna end up in an OTR file with a resource ID like we have with textures, models, DLs etc. So when it comes time to extract that, we just replace this hardcoded string with code to get that string from the resource ID.

As for the fixed location hints, I probably just haven't re-done them yet, or if I have I may not have removed them from this list by mistake, putting back on draft for now.

leggettc18 replied:
Oh, and another benefit to this approach, is that these strings are being constructed in response to a hook. One thing that scripting support will probably come with is a script being able to be registered as a hook callback. This would allow for a translation mod to construct a completely different string if they wanted to, such as if the language needs a completely different sentence structure or something like that. Of course that will likely necessitate bigger changes but that's gonna result in similar amounts of changes no matter where the strings are currently stored.

leggettc18 committed:
Formatting and fixing bad merge conflict resolutions

leggettc18 committed:
Merge branch 'feat/hooks/custom-message' of github.com:leggettc18/Shipwright into feat/hooks/custom-message

leggettc18 committed:
post-merge format run

leggettc18 committed:
Merge branch 'develop' into feat/hooks/custom-message

leggettc18 committed:
Merge branch 'develop' of github.com:HarbourMasters/Shipwright into feat/hooks/custom-message

leggettc18 committed:
Fix post-merge build errros

leggettc18 committed:
Adds colors to more items.

Not sure if I forgot these or if they got lost in a merge, but this adds colors to things like the dungeon keys, maps, and compasses.

leggettc18 committed:
Fix the Yes/No spacing on a couple of messages.

leggettc18 committed:
Merge branch 'develop' into feat/hooks/custom-message

leggettc18 commented:
I got this up to date with latest dev. A few more things to check before I'm ready to remove do not merge from this though.

leggettc18 committed:
clang-format

leggettc18 committed:
Merge branch 'develop' into feat/hooks/custom-message

leggettc18 committed:
Fix mysterious item names in Shops

leggettc18 committed:
Adds articles to gossip stone hints and fixes Biggoron Sword articles.

leggettc18 commented:
I removed Do Not Merge now that this is up to date with latest dev and I did a playthrough of it and couldn't find anything broken. Hinted shop items are a little weird since they show up as something like "They say a child shopkeeper sells Buy Hylian Shield", but that's the only blemish I've found and I don't really know of much of a way around it, other than maybe renaming them to something like "Hylian Shield (Buy)" or something like that so it still grammatically works and is still clear it's referring to the shop item.

serprex commented on soh/soh/Enhancements/randomizer/Messages/EntranceHints.cpp:
```diff
@@ -0,0 +1,168 @@
+}
+
+#define ENTRANCES_SHUFFLED \
+    IS_RANDO&& OTRGlobals::Instance->gRandoContext->GetOption(RSK_SHUFFLE_ENTRANCES).Is(RO_GENERIC_ON)
```
```suggestion
    IS_RANDO && OTRGlobals::Instance->gRandoContext->GetOption(RSK_SHUFFLE_ENTRANCES).Is(RO_GENERIC_ON)
```

leggettc18 replied:
Weird, I thought I remembered clang-format fixing that... oh well, I do need to port over the checkbox for the entrance hints and make that cvar part of the conditional anyway, forgot I had that merged into dev-copper already.

leggettc18 replied:
For some reason, clang-format keeps removing the space. No clue why that's happening, but I'm probably going to turn it off via comments just for that little block.

leggettc18 replied:
Fixed in 39730ad

serprex commented:
@PurpleHato may have thoughts here

re shop hints, why can't string interpolation remove `"Buy "` prefix?

leggettc18 commented:
Hmm, would need to check the other languages to see if it would be that simple, but it's certainly worth a look.

leggettc18 committed:
Merge branch 'develop' into feat/hooks/custom-message

leggettc18 committed:
Fix Entrance Hints and port toggle from dev-copper

leggettc18 committed:
Add articles and remove "Buy" prefix from hinted shop items

leggettc18 commented:
Hints for shop items should be better now. Need to go back over current dev branch to make sure I'm not missing some hint types/changes that may or may not have occurred since when I first made this.

serprex commented on soh/soh/Enhancements/game-interactor/GameInteractor_Hooks.cpp:
```diff
@@ -379,6 +379,14 @@ void GameInteractor_ExecuteOnKaleidoUpdate() {
 
+// MARK: Messages
+void GameInteractor_ExecuteOnOpenText(uint16_t* textId, bool* loadFromMessageTable) {
+    LUSLOG_DEBUG("OnOpenText: textId: {}", *textId);
```
```suggestion
```
is this needed? assuming existing text id print out still there

serprex replied:
text is printing out as `{}` & I confirmed this log line is redundant, going to apply  l go ahead with merge

serprex committed:
remove unnecessary log

serprex commented on soh/soh/Enhancements/randomizer/Messages/Rupees.cpp:
```diff
@@ -0,0 +1,255 @@
+                                              "Diamonds",
+                                              "DNA",
+                                              "Doge",
+                                              "Dogecoin",
```
no to doge, #5488

see also #4817

leggettc18 replied:
I think those lists I can just copy-paste back from develop, will have that fixed today.

leggettc18 replied:
What am I missing from #4817? Or are you just mentioning it to make sure I've seen it? I can't seem to find anything there that I don't already have.

serprex replied:
Yes, just mentioning. Thanks for checking

leggettc18 replied:
Fixed in 73181b3 I believe

leggettc18 committed:
Update rupee names with latest develop

JordanLongstaff commented on soh/soh/Enhancements/InjectItemCounts.cpp:
```diff
@@ -0,0 +1,80 @@
+                      "Vous obtenez un %rQuart de&Coeur%w! Vous en avez collecté&%r[[heartPieceCount]]%w en tout!",
+                      TEXTBOX_TYPE_BLUE);
+    msg.Replace("[[heartPieceCount]]", std::to_string(gSaveContext.ship.stats.heartPieces + 1));
+    msg.AutoFormat(ITEM_HEART_PIECE);
+    msg.LoadIntoFont();
+    *loadFromMessageTable = false;
+}
+
+void InjectItemCounts_Register() {
+    COND_ID_HOOK(OnOpenText, TEXT_GS_FREEZE, CVarGetInteger(CVAR_ENHANCEMENT("InjectItemCounts.GoldSkulltula"), 0),
+                 BuildSkulltulaMessage);
+    COND_ID_HOOK(OnOpenText, TEXT_GS_NO_FREEZE, CVarGetInteger(CVAR_ENHANCEMENT("InjectItemCounts.GoldSkulltula"), 0),
+                 BuildSkulltulaMessage);
+    COND_ID_HOOK(OnOpenText, TEXT_HEART_CONTAINER,
+                 CVarGetInteger(CVAR_ENHANCEMENT("InjectItemCounts.HeartContainer"), 0), BuildHeartContainerMessage);
+    // Heart Pieces don't have documented text IDs after the first one, but
+    // there are 3 more in between TEXT_HEART_PIECE and TEXT_HEART_CONTAINER.
+    COND_ID_HOOK(OnOpenText, TEXT_HEART_PIECE, CVarGetInteger(CVAR_ENHANCEMENT("InjectItemCounts.HeartPiece"), 0),
+                 BuildHeartPieceMessage);
+    COND_ID_HOOK(OnOpenText, TEXT_HEART_PIECE + 1, CVarGetInteger(CVAR_ENHANCEMENT("InjectItemCounts.HeartPiece"), 0),
+                 BuildHeartPieceMessage);
+    COND_ID_HOOK(OnOpenText, TEXT_HEART_PIECE + 2, CVarGetInteger(CVAR_ENHANCEMENT("InjectItemCounts.HeartPiece"), 0),
+                 BuildHeartPieceMessage);
+    COND_ID_HOOK(OnOpenText, TEXT_HEART_PIECE + 3, CVarGetInteger(CVAR_ENHANCEMENT("InjectItemCounts.HeartPiece"), 0),
+                 BuildHeartPieceMessage);
```
*shudder* So many duplications. Please to make some macros.

leggettc18 replied:
I'm gonna respectfully disagree here, though putting the Heart Piece hook registrations in a loop might be reasonable. Most things I could do to remove duplication here would only serve to add more layers of translation someone reading this in the future would have to jump through. I'm just not seeing the benefit.

Malkierian replied:
I will have to agree with leggett here. The nature of the text replacement doesn't really allow simplification of the hook processing itself. Looping is as far as I'd want to take duplication removal.

serprex replied:
@leggettc18 I don't think loop works, since `COND_ID_HOOK` creates static `hookId` variable which will be aliased over each iteration iiuc

fine with the unrolled loop version, no macro

leggettc18 replied:
Ah good catch, you're probably right. Since it's only 4 iterations unrolling the loop is fine.

leggettc18 replied:
Fixed in 1c7bf02

JordanLongstaff commented on soh/soh/Enhancements/InjectItemCounts.cpp:
```diff
@@ -0,0 +1,80 @@
+
+static RegisterShipInitFunc initFunc(InjectItemCounts_Register, { CVAR_ENHANCEMENT("InjectItemCounts.GoldSkulltula"),
+                                                                  CVAR_ENHANCEMENT("InjectItemCounts.HeartContainer"),
+                                                                  CVAR_ENHANCEMENT("InjectItemCounts.HeartPiece") });
```
I prefer to split `initFunc`s like these by which CVar controls them, since there are no crossovers.

leggettc18 replied:
This is reasonable, will do.

leggettc18 replied:
Fixed in 9ede189

leggettc18 committed:
Split InjectItemCounts hook registration funcs

leggettc18 committed:
Merge branch 'feat/hooks/custom-message' of github.com:leggettc18/Shipwright into feat/hooks/custom-message

leggettc18 committed:
clang-format

Pepper0ni commented on soh/soh/Enhancements/randomizer/Messages/Rupees.cpp:
```diff
@@ -0,0 +1,381 @@
+                                             "Diamanten",
+                                             "Dinar",
+                                             "Diridari",
+                                             "Dogecoin",
```
Missed a doge.

leggettc18 replied:
Ah, looks like this one was missed when the PR on develop happened earlier, this is still in latest develop as of this morning when I copy-pasted it. I will get it removed shortly.

leggettc18 replied:
Fixed in fbc4967

leggettc18 committed:
Unroll Heart Piece Count hook loop to prevent overwriting static hookId variables

Pepper0ni reviewed (changes requested):
Only 1 oversight to patch then I'm fine with this.

leggettc18 committed:
Remove leftover Dogecoin reference
